### DISCUSSION TITLE
Use hook or API rules to enforce batch update

### DISCUSSION THREAD

**Discussion by moritz4** [pocketbase/pocketbase#5998] [Q&A]
I have a use case where some of my records in a collection are logically grouped together. I want to enforce that if a user updates a certain field in a record, they are also updating the other records which are grouped together. I am using the new JS SDK BatchService on the client. How can I use hooks or API rules to enforce this? 

I can use the `onBatchRequest` to make sure that all grouped records are being updated with the same value:

```js
onBatchRequest((e) => {
	// Get ids of grouped records
	const ids = getGroupedRecords();

	// Throw error if not all records are being updated at the same time
	if (not all e.batch.url ends with ids) { throw new Error}

	// Throw error if not all records are being updated with the same value
	if (not all e.batch.method === 'PATCH') { throw new Error}
	if (not all e.batch.body is the same) { throw new Error}

	e.next();
});
```

However, this does not prevent someone from updating a single record in that collection. A batch update will also trigger a `onRecordUpdateRequest` hook for each question being updated, but I don't see any obvious way to check in that hook that this a batch request.

**Comment by chnfyi** [comment]
From my understanding of what you are trying to do, you could do it in a simpler way having a `group` field on every records of your collection.

**Not tested example**

```javascript
const updatedRecord = await pb.collection("your_collection").getOne("record_id")

// Retrieve records ids from the same group
const records = await pb.collection("your_collection").getList(1, 50, {
    filter: pb.filter("group = {:group1}", {group1: updatedRecord.group}),
    fields: "id"
});

const updatedFields = { "field1": updatedRecord.field1, "field2": updatedRecord.field2 }

// Create a new batch instance
const batch = pb.createBatch()

// Register all update requests
for (let record of records) {
    batch.collection("your_collection").update(record.id, updatedFields)
}

// Send the batch request
const result = await batch.send()
``` 

_Note: To achieve it with SDK you need to modify your API rules in order to allow others to view and update. Otherwise, if you want to keep more restrictive API rules you can register a custom route where you send your update request and update all records that belong to the same group._

**Reply by moritz4** [reply]
Thank you, I think the problem still remains, which is that a client can update just one of these grouped records, when I am trying to force them to update all records in the group at the same time. E.g.

```js
// This should not be allowed and should be blocked by hooks or API rules because we are only updating a single record in the group
pb.collection("your_collection").update("record_1", {title: "The Maltese Falcon"});

// We have not updated record_2 or record_3
```

**Reply by chnfyi** [reply]
You can modify your collection API rules to allow only superusers to update records. Then register a custom route where you send your update requests and where you update all records that belong to the same group.

**Not tested example**

In browser
```javascript
await pb.send("/api/myapp/updateGroupOfRecords", {
    headers: { "Content-Type": "application/json" },
    body: { "field1": newValue1, "field2": newValue2 }
})
```

In pb_hooks/routes.pb.js
```javascript
routerAdd("POST", "/api/myapp/updateGroupOfRecords", (e) => {
    const info = e.requestInfo()
    const record = info.auth
    const body = info.body
    const fields = Object.entries(body)
    const recordGroup = record.get("group")
    const records = arrayOf(new Record())
    
    $app.recordQuery("your_collection")
        .andWhere($dbx.exp("group = {:group}", { group: recordGroup }))
        .all(records)

    for (let record of records) {
        for (let [k, v] of fields) {
            record.set(k, v)
        }
    }
    
    return e.noContent(204)
}, $apis.requireAuth())
```
